core/node/node_framework/src/implementations/layers/web3_api/tx_sink/whitelisted_master_pool_sink_layer.rs

```diff
@@ -0,0 +1,183 @@
+use serde::Deserialize;
+use tokio::time::interval;
+use zksync_config::configs::api::DeploymentAllowlist;
+use zksync_node_api_server::tx_sender::{
+    master_pool_sink::MasterPoolSink, shared_allow_list::SharedAllowList,
+    whitelisted_deploy_pool_sink::WhitelistedDeployPoolSink,
+};
+use zksync_types::Address;
+
+use crate::{
+    implementations::resources::{
+        pools::{MasterPool, PoolResource},
+        web3_api::TxSinkResource,
+    },
+    service::StopReceiver,
+    task::{Task, TaskId, TaskKind},
+    wiring_layer::{WiringError, WiringLayer},
+    FromContext, IntoContext,
+};
+
+/// Wiring layer for [`WhitelistedDeployPoolSink`] that wraps a `MasterPoolSink` and enables allowlist filtering.
+pub struct WhitelistedMasterPoolSinkLayer {
+    pub deployment_allowlist: DeploymentAllowlist,
+}
+
+#[derive(Debug, FromContext)]
+#[context(crate = crate)]
+pub struct Input {
+    pub master_pool: PoolResource<MasterPool>,
+}
+
+#[derive(Debug, IntoContext)]
+#[context(crate = crate)]
+pub struct Output {
+    pub tx_sink: TxSinkResource,
+    #[context(task)]
+    pub allow_list_task: AllowListTask,
+}
+
+#[async_trait]
+impl WiringLayer for WhitelistedMasterPoolSinkLayer {
+    type Input = Input;
+    type Output = Output;
+
+    fn layer_name(&self) -> &'static str {
+        "whitelisted_master_pool_sink_layer"
+    }
+
+    async fn wire(self, input: Self::Input) -> Result<Self::Output, WiringError> {
+        let pool = input.master_pool.get().await?;
+        let master_pool_sink = MasterPoolSink::new(pool);
+
+        let shared_allowlist = SharedAllowList::new();
+
+        let allow_list_task = AllowListTask::new(
+            self.deployment_allowlist
+                .http_file_url()
+                .expect("DeploymentAllowlist must contain a URL")
+                .to_string(),
+            self.deployment_allowlist.refresh_interval(),
+            shared_allowlist.clone(),
+        );
+
+        let tx_sink = WhitelistedDeployPoolSink::new(master_pool_sink, shared_allowlist).into();
+
+        Ok(Output {
+            tx_sink,
+            allow_list_task,
+        })
+    }
+}
+
+/// Task that periodically fetches and updates the allowlist from a remote HTTP source.
+#[derive(Debug)]
+pub struct AllowListTask {
+    url: String,
+    refresh_interval: Duration,
+    allowlist: SharedAllowList,
+    client: Client,
+    etag: tokio::sync::Mutex<Option<String>>,
+}
+
+#[derive(Debug, Deserialize)]
+struct WhitelistResponse {
+    addresses: Vec<Address>,
+}
+
+impl AllowListTask {
+    const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+    pub fn new(url: String, refresh_interval: Duration, allowlist: SharedAllowList) -> Self {
+        Self {
+            url,
+            refresh_interval,
+            allowlist,
+            client: Client::new(),
+            etag: tokio::sync::Mutex::new(None),
+        }
+    }
+
+    async fn fetch(&self) -> anyhow::Result<Option<HashSet<Address>>> {
```

Comment:
You can use `&mut self` receiver here to avoid wrapping the `ETag` field into a `Mutex`. Alternatively, return a tag together with data, so that it's updated on the caller side (IMO, that would be a preferable option).